File: guilt/commands/batch.py
from guilt.interfaces.command import CommandInterface
from guilt.interfaces.services.repository_factory import RepositoryFactoryServiceInterface
from guilt.interfaces.services.user import UserServiceInterface
from guilt.interfaces.services.carbon_intensity_forecast import CarbonIntensityForecastServiceInterface
from guilt.interfaces.services.ip_info import IpInfoServiceInterface
from guilt.utility.time_series_data import WindowWithLowestSumResult
from guilt.utility.format_duration import format_duration
from guilt.models.unprocessed_job import UnprocessedJob
from guilt.mappers import map_to
from guilt.utility import slurm_batch
from guilt.utility.calculate_tdp_per_core import calculate_tdp_per_core
from argparse import ArgumentParser, Namespace
from pathlib import Path
from datetime import datetime, timedelta
from typing import Optional
from dataclasses import replace

class BatchCommand(CommandInterface):
  def __init__(
    self,
    repository_factory_service: RepositoryFactoryServiceInterface,
    user_service: UserServiceInterface,
    carbon_intensity_forecast_service: CarbonIntensityForecastServiceInterface,
    ip_info_service: IpInfoServiceInterface
  ) -> None:
    self._repository_factory_service = repository_factory_service
    self._user_service = user_service
    self._carbon_intensity_forecast_service = carbon_intensity_forecast_service
    self._ip_info_service = ip_info_service

  @staticmethod
  def name() -> str:
    return "batch"

  @staticmethod
  def configure_subparser(subparser: ArgumentParser) -> None:
    subparser.add_argument(
      "input",
      help="Input file or argument for batch command"
    )

  def execute(self, args: Namespace) -> None:
    path = Path(args.input)

    with path.open("r") as file:
      content = file.read().splitlines()

    current_user = self._user_service.get_current_user()

    if current_user is None:
      raise RuntimeError("No user is currently logged in. Please log in to continue.")

    guilt_directives = map_to.guilt_script_directives.from_json_directives(
      map_to.json.from_directive_lines(content, "#GUILT"),
      self._repository_factory_service.get_cpu_profiles_repository(current_user)
    )

    slurm_directives = map_to.slurm_script_directives.from_json_directives(
      map_to.json.from_directive_lines(content, "#SBATCH")
    )

    test_job = slurm_batch.test(path, None)
    test_job

    earliest_possible_start_time = test_job.start_time
    latest_forecast_time = datetime.now() + timedelta(days=2, minutes=-30)

    start_time: Optional[datetime] = None
    if earliest_possible_start_time + slurm_directives.time > latest_forecast_time:
      print("Your job is extends into unforecasted carbon intensity data. No delays will be applied.")
    else:
      intensity_data = map_to.time_series_data.from_carbon_intensity_forecast_result(
        self._carbon_intensity_forecast_service.get_forecast(
          earliest_possible_start_time,
          latest_forecast_time,
          self._ip_info_service.get_ip_info().postal
        )
      )

      tdp = slurm_directives.tasks_per_node * slurm_directives.cpus_per_task * slurm_directives.nodes * calculate_tdp_per_core(guilt_directives.cpu_profile)

      earliest_possible_time_sum = intensity_data.get_window_sum(
        earliest_possible_start_time,
        earliest_possible_start_time + slurm_directives.time
      )
      earliest_possible_time_grams = (tdp / 1000) * (earliest_possible_time_sum / 3600)

      best_times = intensity_data.get_windows_with_lowest_sum(
        earliest_possible_start_time,
        intensity_data.get_last_time() - slurm_directives.time - timedelta(minutes=1),
        slurm_directives.time,
        timedelta(minutes=1) 
      )

      best_times_with_grams: list[WindowWithLowestSumResult] = []
      for best_time in best_times:
        emissions = (tdp / 1000) * (best_time.sum_value / 3600)
        if emissions > earliest_possible_time_grams:
          continue

        emissions_saved = earliest_possible_time_grams - emissions

        delay = best_time.start_time - earliest_possible_start_time
        delay_seconds = delay.total_seconds() or 1

        if emissions_saved / delay_seconds > 0.001:
          best_times_with_grams.append(replace(best_time, sum_value=emissions))

      if not best_times_with_grams:
        print("No suitable alternative start times for the job could be found. No delays will be applied.")
      else:
        start_time = best_times_with_grams[0].start_time

        delay_seconds = (start_time - earliest_possible_start_time).total_seconds()

        print(f"If you had used 'sbatch', the job would have begun at {earliest_possible_start_time.strftime('%Y-%m-%d %H:%M')} and emitted {earliest_possible_time_grams:.2f} grams of CO2.")
        print(f"By delaying the job to {best_times_with_grams[0].start_time.strftime('%Y-%m-%d %H:%M')}, it will emit {best_times_with_grams[0].sum_value:.2f} grams of CO2.")
        print(f"You are saving {earliest_possible_time_grams- best_times_with_grams[0].sum_value:.2f} grams of CO2 by delaying the job by {format_duration(delay_seconds)}.")

    job_id = slurm_batch.submit(path, start_time)
    
    unprocessed_jobs_repository = self._repository_factory_service.get_unprocessed_jobs_repository(current_user)

    if job_id in [job.job_id for job in unprocessed_jobs_repository.get_all()]:
      print(f"Unprocessed job with ID {job_id} already exists.")
      return
    
    unprocessed_jobs_repository.upsert(UnprocessedJob(
      job_id,
      guilt_directives.cpu_profile
    ))

    unprocessed_jobs_repository.save()

    print(f"Job submitted with ID {job_id}.")

File: src/commands/batch.rs
use crate::carbon_intensity::{CarbonIntensityAggregator, api::ApiFetchCarbonIntensity};
use crate::guilt_directory::GuiltDirectoryManager;
use crate::ip_info::fetch_ip_info;
use crate::models::UnprocessedJob;
use crate::script_directives::{GuiltScriptDirectives, SlurmScriptDirectives};
use crate::slurm::batch;
use crate::users::get_current_user;
use chrono::{DateTime, Duration, Utc};
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum BatchCommandError {
    #[error("CPU profile '{0}' not found.")]
    CpuProfileNotFound(String),
}

fn resolve_path(path_str: &str) -> std::io::Result<PathBuf> {
    let path = Path::new(path_str);

    if path.is_absolute() {
        Ok(path.to_path_buf())
    } else {
        let cwd = env::current_dir()?;
        Ok(cwd.join(path))
    }
}

static SEARCH_RESOLUTION: i64 = 15;

pub async fn run(job: &str) -> anyhow::Result<()> {
    let path = resolve_path(job)?;

    let contents = fs::read_to_string(&path)?;

    let guilt_directives = GuiltScriptDirectives::from_file_contents(&contents)?;
    let slurm_directives = SlurmScriptDirectives::from_file_contents(&contents)?;

    let test_job = batch::test(job, None)?;

    let earliest_possible_start_time = test_job.start_time;
    let latest_forecast_time = Utc::now() + Duration::days(2) - Duration::minutes(30);

    let mut guilt_dir_manager = GuiltDirectoryManager::read_for_user(&get_current_user()?);
    let cpu_profile = guilt_dir_manager
        .get_cpu_profile(&guilt_directives.cpu_profile)?
        .ok_or(BatchCommandError::CpuProfileNotFound(
            guilt_directives.cpu_profile,
        ))?;

    let begin: Option<DateTime<Utc>> = if earliest_possible_start_time + slurm_directives.time
        > latest_forecast_time
    {
        println!(
            "Your job is extends into unforecasted carbon intensity data. No delays will be applied."
        );
        None
    } else {
        let tdp_kilowatts = (slurm_directives.tasks_per_node
            * slurm_directives.cpus_per_task
            * slurm_directives.nodes) as f64
            * (cpu_profile.tdp as f64 / cpu_profile.cores as f64)
            / 1000.0;

        let ip_info = fetch_ip_info().await?;
        let mut aggregator =
            CarbonIntensityAggregator::new(ApiFetchCarbonIntensity::new(ip_info.postal));

        // initialise the cache (make all requests at start, instead of many small requests later)
        aggregator
            .get_segments(earliest_possible_start_time, latest_forecast_time)
            .await?;

        let earliest_possible_intensity = aggregator
            .get_average_intensity(
                earliest_possible_start_time,
                earliest_possible_start_time + slurm_directives.time,
            )
            .await?;
        let earliest_possible_emissions = earliest_possible_intensity * tdp_kilowatts;

        let latest_possible_start_time = latest_forecast_time - slurm_directives.time;
        let start_times = (1..(latest_possible_start_time - earliest_possible_start_time)
            .num_minutes()
            / SEARCH_RESOLUTION)
            .map(|m| earliest_possible_start_time + Duration::minutes(m * SEARCH_RESOLUTION));
        let time_windows = start_times.map(|start| (start, start + slurm_directives.time));

        let mut intensities: Vec<(DateTime<Utc>, f64)> = Vec::new();
        for (start, end) in time_windows {
            let intensity = aggregator
                .get_average_intensity(start, end)
                .await
                .unwrap_or(f64::MAX);
            intensities.push((start, intensity));
        }
        intensities.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal));

        let best_time = intensities.iter().find_map(|(start, intensity)| {
            if intensity > &earliest_possible_intensity {
                None
            } else {
                let emissions = tdp_kilowatts * intensity;

                let emissions_reduction = earliest_possible_emissions - emissions;
                let delay = *start - earliest_possible_start_time;
                let delay_seconds = delay.num_seconds().max(1) as f64;

                if emissions_reduction / delay_seconds > 0.001 {
                    Some((*start, emissions))
                } else {
                    None
                }
            }
        });

        match best_time {
            Some((start, emissions)) => {
                println!(
                    "If you had used 'sbatch', the job would have begun at {} and emitted {:.2} grams of CO2.",
                    earliest_possible_start_time.format("%Y-%m-%d %H:%M"),
                    earliest_possible_emissions
                );
                println!(
                    "By delaying the job to {}, it will emit {:.2} grams of CO2.",
                    start.format("%Y-%m-%d %H:%M"),
                    emissions
                );
                println!(
                    "You are saving {:.2} grams of CO2 by delaying the job.",
                    earliest_possible_emissions - emissions
                );

                Some(start)
            }
            None => {
                println!(
                    "Could not find a better time to run your job, no delays will be applied."
                );
                None
            }
        }
    };

    let local_begin = begin.map(|b| b.with_timezone(&chrono::Local));
    let naive_begin = local_begin.map(|b| b.naive_local());

    let job_id = batch::submit(job, naive_begin)?;
    println!("Submitted job {}.", job_id);

    let unprocessed_job = UnprocessedJob {
        job_id,
        cpu_profile,
    };

    guilt_dir_manager.upsert_unprocessed_job(unprocessed_job)?;
    guilt_dir_manager.write()?;

    Ok(())
}
